### ISSUE TITLE
Query Diffing Clarification

### ISSUE THREAD

**Issue by nickhudkins** [facebook/relay#1197]
I noticed some behavior that, at first glance seems unexpected, but very well could be as designed.

**tl;dr:** Lots (_n_, where _n_ is number of nodes in connection) node queries are created to fulfill data needs for nodes in a connection.

See here for a video you can watch of what occurs: http://n.hudk.in/gKx7

and see here:
https://boiling-lowlands-71549.herokuapp.com/ for a deployed version of the relay TODO example modified to exhibit said unexpected (though perhaps correct?) behavior.

The relevant RelayContainer fragment definitions can be seen here:
```javascript

/* ... TodoList.js ... */
  fragments: {
    viewer: () => Relay.QL`
      fragment on User {
        todos(status: $status, first: 2147483647) {
          edges {
            node {
              id,
              ${Todo.getFragment('todo')},
            },
          },
        },
      }
    `,
  },
  /* ... */
```

```javascript
/* ... TodoListExtra.js ... */
  fragments: {
    viewer: () => Relay.QL`
      fragment on User {
        completedCount,
        todos(status: $status, first: 2147483647) {
          edges {
            node {
              id,
              extra
              ${Todo.getFragment('todo')},
            },
          },
        },
      }
    `,
  },
  /* ... */
```

You can reproduce this, by adding several Todos then clicking "Extra Fields" in the footer which will route to /complete/extra (and render TodoListExtra), TodoListExtra's data needs will be satisfied by the Relay Store, except for the additional field titled: `extra` at which point **n** node queries will be made to fulfill the `extra` field for the `todo` nodes in the `todos` connection.

This seems correct as far as ensuring proper data is fetched, but could result in some VERY nasty requests with larger data sets. 

Is there a better way to get some extra data about nodes you have already fetched without so many node queries?

An optimization (in my head) would be something along the lines of:

` nodes(ids: $listOfNodeIds) { extra }` 

Looking forward to hearing if others have experienced this, have solved it, or if it's something that is worth considering further as an improvement to relay.

**Comment by NevilleS** [comment]
I'm also interested in this... Nick and I discussed this a bit and it seems like a dangerous "gotcha!" that requesting additional fields on a connection fragment would result in N queries instead of just a single fetch. My guess is that this is done to avoid refetching the connection (and therefore potentially changing which edges exist), but that tradeoff might not actually be what you want, especially if N is large!

**Comment by nickhudkins** [comment]
Awesome. thank you! I wish I had searched better :)

**Comment by josephsavona** [comment]
All good!
